// @cedra-labs/wallet-adapter-plugin/src/CedraConnectAppleWallet.ts
// Copyright © Cedra

// SPDX-License-Identifier: Apache-2.0

import {
  CedraConnectNamespace,
  CedraDisconnectNamespace,
  CedraFeatures,
  CedraGetAccountNamespace,
  CedraGetNetworkNamespace,
  CedraOnAccountChangeNamespace,
  CedraOnNetworkChangeNamespace,
  CedraSignAndSubmitTransactionFeature,
  CedraSignAndSubmitTransactionNamespace,
  CedraSignInNamespace,
  CedraSignMessageNamespace,
  CedraSignTransactionNamespace,
  CedraWallet,
} from "@cedra-labs/wallet-standard";

import {
  CedraConnectWallet,
  CedraConnectWalletConfig,
} from "./CedraConnectWallet";

import { walletMetadata } from "./config";

export class CedraConnectAppleWallet
  extends CedraConnectWallet
  implements CedraWallet
{
  readonly name = walletMetadata.apple.name;

  readonly icon = walletMetadata.apple.icon;

  readonly url = walletMetadata.apple.url;

  constructor(config: CedraConnectWalletConfig) {
    super({ ...config, provider: "apple" });
  }

  // region CedraWallet

  get features(): CedraFeatures & CedraSignAndSubmitTransactionFeature {
    return {
      [CedraConnectNamespace]: {
        connect: this.connect.bind(this),

        version: "1.0.0",
      },

      [CedraDisconnectNamespace]: {
        disconnect: this.disconnect.bind(this),

        version: "1.0.0",
      },

      [CedraGetAccountNamespace]: {
        account: this.getAccount.bind(this),

        version: "1.0.0",
      },

      [CedraGetNetworkNamespace]: {
        network: this.getNetwork.bind(this),

        version: "1.0.0",
      },

      [CedraOnAccountChangeNamespace]: {
        onAccountChange: this.onAccountChange.bind(this),

        version: "1.0.0",
      },

      [CedraOnNetworkChangeNamespace]: {
        onNetworkChange: this.onNetworkChange.bind(this),

        version: "1.0.0",
      },

      [CedraSignAndSubmitTransactionNamespace]: {
        signAndSubmitTransaction: this.signAndSubmitTransaction.bind(this),

        version: "1.1.0",
      },

      [CedraSignInNamespace]: {
        signIn: this.signIn.bind(this),

        version: "0.1.0",
      },

      [CedraSignMessageNamespace]: {
        signMessage: this.signMessage.bind(this),

        version: "1.0.0",
      },

      [CedraSignTransactionNamespace]: {
        signTransaction: this.signTransaction.bind(this),

        version: "1.1",
      },
    };
  }

  // endregion
}
