// @cedra-labs/identity-wallet-api/src/serialization/bcsSerialization.ts
// Copyright © Cedra
// SPDX-License-Identifier: Apache-2.0

import {
  Deserializer,
  Hex,
  Serializer,
  Serializable,
} from "@cedra-labs/ts-sdk";

interface BcsSerializable {
  bcsToBytes(): Uint8Array;
  bcsToHex(): Hex;
}

export type DeserializableClass<T extends Serializable> = {
  deserialize(deserializer: Deserializer): T;
};

function isSerializable(value: any): value is Serializable {
  return (
    (value as Serializable)?.serialize !== undefined &&
    (value as Serializable)?.bcsToBytes !== undefined &&
    (value as Serializable)?.bcsToHex !== undefined
  );
}

/**
 * Check if a value is BCS serializable
 */
export function isBcsSerializable(value: any): value is Serializable {
  return isSerializable(value);
}

export function bcsSerialize(serializable: BcsSerializable) {
  if (isSerializable(serializable)) {
    return serializable.bcsToHex().toString();
  }
  const serializedValueBytes = serializable.bcsToBytes();
  return Hex.fromHexInput(serializedValueBytes).toString();
}

export function bcsDeserialize<T extends Serializable>(
  deserializableClass: DeserializableClass<T>,
  serializedValue: string,
) {
  const serializedValueBytes =
    Hex.fromHexString(serializedValue).toUint8Array();
  const deserializer = new Deserializer(serializedValueBytes);
  return deserializableClass.deserialize(deserializer);
}
