// @cedra-labs/wallet-api/src/shared/CedraSignInOutput.ts
// Copyright © Cedra
// SPDX-License-Identifier: Apache-2.0

import { Deserializer, Serializer } from '@cedra-labs/ts-sdk';
import { AccountInfo, CedraSignInOutput } from '@cedra-labs/wallet-standard';
import { deserializeSignature, serializeSignature } from './Signature';
import { serializeAccountInfo, deserializeAccountInfo } from './AccountInfo';

export function serializeCedraSignInOutput(serializer: Serializer, value: CedraSignInOutput) {
  serializeAccountInfo(serializer, value.account);
  serializer.serializeStr(JSON.stringify(value.input));
  serializer.serializeStr(value.plainText);
  serializeSignature(serializer, value.signature);
  serializer.serializeBytes(value.signingMessage);
  serializer.serializeStr(value.type);
}

export function deserializeCedraSignInOutput(deserializer: Deserializer): CedraSignInOutput {
  return {
    account: new AccountInfo(deserializeAccountInfo(deserializer)),
    input: JSON.parse(deserializer.deserializeStr()),
    plainText: deserializer.deserializeStr(),
    signature: deserializeSignature(deserializer),
    signingMessage: deserializer.deserializeBytes(),
    type: deserializer.deserializeStr(),
  };
}
